File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        map = {'(':')' , '[':']' , '{' : '}' }

        for i in s:
            if i in map:
                stack.append(i)

            elif len(stack) == 0 or map[stack.pop()] != i:
                return False

        return len(stack) == 0
                

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();

        for (char c : s.toCharArray()) {
            if (c == '[' || c == '{' || c == '(') {
                st.push(c);
            } else if (st.isEmpty()) {
                return false;
            } else if ((c == ')' && st.peek() == '(') ||
                    (c == ']' && st.peek() == '[') ||
                    (c == '}' && st.peek() == '{')) {
                st.pop();
            } else {
                return false;
            }
        }
        return st.isEmpty();
    }
}